### PR TITLE
ADEN-6972 Add geo recognition with sampling support

### PR DESCRIPTION
**Pull request by v3rth**

### CODE REVIEW

File: src/utils/random.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
+// TODO remove this module
```
Comment (fraszczakszymon) [inline src/utils/random.js:1]: Can't we use sampler.js?

File: src/utils/geo.js [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+
+function isCountryWithSampling(geo) {
+	return value => !value.startsWith(negativePrefix) && value.includes(geo + samplingChar);
+}
+
+function getSamplingLimits(value) {
+	return parseFloat(value.split(samplingChar)[1]) / 100;
+}
+
+function getResult(s, name) {
```
Comment (fraszczakszymon) [inline src/utils/geo.js]: Rename `s` please

File: src/utils/geo.js [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+
+const earth = 'XX',
+	negativePrefix = 'non-',
+	samplingChar = '/';
+
+let geoData = null,
+	cache = [];
+
+function isCountryWithSampling(geo) {
+	return value => !value.startsWith(negativePrefix) && value.includes(geo + samplingChar);
```
Comment (fraszczakszymon) [inline src/utils/geo.js]: What about `non-US/50`? Are we going to support it?

File: src/utils/geo.js [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+import Random from './random';
+
+const earth = 'XX',
+	negativePrefix = 'non-',
+	samplingChar = '/';
+
+let geoData = null,
+	cache = [];
+
+function isCountryWithSampling(geo) {
```
Comment (fraszczakszymon) [inline src/utils/geo.js]: It isn't really country, right? Can we rename it to `hasSampling`? Or something similar

File: src/utils/geo.js [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+			group: result ? 'B' : 'A',
+			limit: result ? limit : 100 - limit,
+			result: result
+		};
+	}
+
+	return result;
+}
+
+function sampleGeo(countryList, geo, name) {
```
Comment (fraszczakszymon) [inline src/utils/geo.js]: s/sampleGeo/isSampledGeo
what do you think?

File: src/utils/geo.js (hunk 3 above)
Comment (fraszczakszymon) [reply src/utils/geo.js]: What about IE11 (includes)? 😈

File: src/utils/geo.js [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+function getTrackingLog (e) {
+	let obj = cache[e];
+	return `${obj.name}_${obj.group}_${obj.limit}`;
+}
+
+export function resetCache() {
+	cache = {};
+}
+
+export function getTrackingValues() {
```
Comment (fraszczakszymon) [inline src/utils/geo.js]: s/getTrackingValues/getSamplingResults
?

File: src/utils/geo.js [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+		countryList.indexOf(`${negativePrefix}${earth}-${getContinentCode()}`) > -1
+	);
+}
+
+function getTrackingLog (e) {
+	let obj = cache[e];
+	return `${obj.name}_${obj.group}_${obj.limit}`;
+}
+
+export function resetCache() {
```
Comment (fraszczakszymon) [inline src/utils/geo.js]: s/resetCache/resetSamplingCache
?

File: src/utils/geo.js (hunk 3 above)
Comment (v3rth) [reply src/utils/geo.js]: Nope `non-US/50` isn't supported

File: src/utils/random.js (hunk 1 above)
Comment (v3rth) [reply src/utils/random.js:1]: discussed offline, sampler.js will be removed

**Comment by fraszczakszymon** [comment]
Remember about label (:

File: package.json [hunk 8]
```diff
@@ -10,6 +10,7 @@
     "@wikia/ad-engine": "git+https://github.com/Wikia/ad-engine.git#v10.0.0",
     "core-decorators": "^0.20.0",
     "events": "^1.1.1",
+    "js-cookie": "^2.1.4",
```
Comment (nandy-andy) [inline package.json]: Is it safe? :trollface:

Comment (nandy-andy) [reply package.json]: Snyk says it's safe: [Known Vulnerabilities](https://snyk.io/test/github/wikia/ad-products?targetFile=package.json)

Comment (v3rth) [reply package.json]: I'm not sure, I moved this module from f2 repository, where this package [is used](https://github.com/Wikia/f2/blob/master/frontend/package.json#L32)

File: src/utils/geo.js [hunk 9]
```diff
@@ -0,0 +1,201 @@
+import Cookies from 'js-cookie';
```
Comment (nandy-andy) [inline src/utils/geo.js:1]: Just a thought which doesn't need to be addressed now: how about creating another new repo for geo.js? ;) It's not really an `ad-product` and it also doesn't really fit to `ad-engine`. Poor module :-(

File: src/index.js [hunk 10]
```diff
@@ -11,3 +11,4 @@ set(window, versionField, 'v<?= PACKAGE(version) ?>');
 
 export * from './common';
 export * from './templates';
+export * from './utils';
```
Comment (fraszczakszymon) [inline src/index.js]: Can we reexport it like in ad-engine?

File: spec/utils/geo.spec.js [hunk 11]
```diff
@@ -0,0 +1,323 @@
+import {assert} from 'chai';
+import {resetSamplingCache, getCountryCode, getContinentCode, getRegionCode, isProperGeo, setGeoData, getSamplingResults} from '../../src/utils/geo.js';
+import Random from '../../src/utils/random.js';
+import sinon from 'sinon';
+import {context, slotService} from '@wikia/ad-engine/dist/ad-engine';
```
Comment (leshniak) [inline spec/utils/geo.spec.js]: Just @wikia/ad-engine should work as well

**Pull request merged**